cmd/dagger/flags.go

```diff
@@ -250,16 +250,45 @@ type secretValue struct {
 	plaintext string
 }
 
+const (
+	envSecretSource  = "env"
+	fileSecretSource = "file"
+)
+
 func (v *secretValue) Type() string {
 	return Secret
 }
 
 func (v *secretValue) Set(s string) error {
+	secretSource, val, ok := strings.Cut(s, ":")
+	if !ok {
+		val = secretSource
+		secretSource = envSecretSource
+	}
+
+	switch secretSource {
+	case envSecretSource:
+		plaintext, ok := os.LookupEnv(val)
+		if !ok {
+			return fmt.Errorf("secret env var source not found: %q", val)
+		}
+		v.plaintext = plaintext
+
+	case fileSecretSource:
+		plaintext, err := os.ReadFile(val)
+		if err != nil {
+			return fmt.Errorf("failed to read secret file source: %q", val)
+		}
+		v.plaintext = string(plaintext)
+
+	default:
+		return fmt.Errorf("unsupported secret arg source: %q", secretSource)
+	}
+
 	// NB: If we allow getting the name from the dagger.Secret instance,
 	// it can be vulnerable to brute force attacks.
-	hash := sha256.Sum256([]byte(s))
+	hash := sha256.Sum256([]byte(v.plaintext))
```

Comment:
I think it's better to keep `v.name = s` (or hashed) rather than based on the value now that it's possible  (see NB above).

Reply:
Not immediately following (bear with me as I rubber duck):
1. In the previous implementation
   1. `s` was the actual plaintext of the secret value
   2. `hash` was the sha hash of the plaintext
   3. `v.name` was set to the `hash`
2. In the updated implementation
   1. `hash` is the sha hash of the plaintext
   2. `v.name` is set to the `hash`

What's the difference?

Reply:
No difference now, what I'm saying is that it's possible to be different. Use the value set in the CLI as the name and the extracted value as plaintext. That way the secret name isn't based on the value which is better. It wasn't possible before.

I realize there was an argument in favor of a name based on value making it easier to invalidate the cache but it makes me iffy because if you can get a secret by name, you get try to guess what the secret value is that way.

Reply:
Ah okay sorry I was misunderstanding! I'm following now and agree that makes sense, will update 👍

Reply:
Okay actually I thought about it more and I think it's probably better to leave the secret name as the hash for now. The problem is that if the secret name is just e.g. `env:MY_TOKEN`, now the engine server is storing it in memory with that name in the in-memory secret store for the session.

Each session gets its own isolated secret store (and there's an issue open in linear to lock down further such that each *module* in a session gets its own secret store) but for general "defense in depth" considerations, if you assume that something went wrong and somehow someone was successfully making API calls that hit a secret store they shouldn't be hitting then:
1. If the secret name is a hash, they'd need to guess the hash of the plaintext to lookup a given secret. That's only feasible if the value is insecure in the first place (and they make tons of API calls to brute force, etc.)
2.  If the secret name is just the arg value like `env:TOPSECRET` they don't need to guess anything and can just look it up by that name.

I agree that even making the name be the hash of the value isn't as ideal as it could be, but from that perspective I think it's actually the best at the moment.

That being said, I am adjusting this so that the `String` method on `secretValue` in the CLI is the raw arg value (e.g. `env:TOPSECRET`) rather than the secret name, which at least hides the SHA a tiny bit more (marginal improvement but may as well)

Reply:
Pushed that update https://github.com/dagger/dagger/pull/6314/commits/4fb309b02fd798afd4331397913024aff5427121

Reply:
Sorry, I didn't explain it correctly 🙂 

The problem isn't whether you can get a secret by name, it's getting the name of a secret from its instance:
```python
import base64

import dagger
from dagger import function


@function
async def secret_name(token: dagger.Secret) -> str:
    sid = await token.id()
    enc_val = sid.split(":")[1]
    return base64.b64decode(enc_val).decode()
```

Of course if the secret value is complex enough it's less feasible to brute force, but I'm thinking about the very easy ones (password, 123). Even an empty value would become a static thing you can check for. So I'd be more at ease with zero-knowledge as much as possible. If that name were something "static" like MY_TOKEN then it feels safer to me.

Of course getting the secret name from id isn't necessarily stable and will possibly change.

Reply:
Oh okay sure sure. Yeah I think it would be better if the name passed around was genuinely random and just mapped back to the value in the secret store, but that's hard in the short term because it means you'd never get caching when using secrets (ways to work around it potentially, but get somewhat complicated).

I just think that switching to using e.g. `env:MY_TOKEN` more trades one insecurity for another potential one. The balance is debatable but for the time being it's not clear cut enough to me to change as part of this particular PR.